### ISSUE TITLE
i18n: pass locale to shop-api 

### ISSUE THREAD

**Issue by redaready** [vendure-ecommerce/storefront-qwik-starter#104]
is there a way to pass the locale value of "extractLang(request.headers.get('accept-language'), request.url)" to shop-api call?

i have tried something like this, but a error raised: "Internal server error: Reading `locale` outside of context."
```
const executeRequest = async (options: Options) => {
	const locale = getLocale();
	const httpResponse = await fetch(`${ENV_VARIABLES.VITE_VENDURE_PUBLIC_URL}/?languageCode=${locale}`, options);
	return await extractTokenAndData(httpResponse);
};
```

**Comment by gioboa** [comment]
I will investigate in it

**Comment by redaready** [comment]
with some help of sourcegraph cody:

That error means you are calling getLocale() outside of a Qwik component or route. getLocale() relies on Qwik's locale context, so it can only be called within a Qwik component/route.

Some options to fix this:

Pass the locale as an argument to the function instead of calling getLocale():
```
function doSomething(locale: string) {
  // Use locale here instead of calling getLocale()
}
```
Wrap the call in a Qwik component:
```
component$(() => {
  const locale = getLocale();
  // Use locale here
});
```
Use a Qwik route and access the locale from the request event:
```
export const onRequest: RequestHandler = ({ locale }) => {
  // Use locale here
};
```
Store the locale in a variable outside of the Qwik context and pass that in:
```
let locale;

component$(() => {
  locale = getLocale();
});

function doSomething() {
  // Use locale here instead of calling getLocale()
}
```
Hope this helps resolve your error! Let me know if you have any other questions.

**Comment by redaready** [comment]
@giovramirez i think locale is a little bit diffrent with token, token is generated by vendure and then stored in cookie, locale on other hand, we must pass it to sdk calls at first invocations (ex: useCollectionsLoader and useAvailableCountriesLoader etc)... so i think we have no other choice than pass locale parameter to many sdk calls :-(

**Comment by giovramirez** [comment]
You're right @redaready. Besides, we can't use a Cookie since most of the calls are executed on the server, so it's not an option to get the locale from Cookie.

We have this [interceptor](https://github.com/vendure-ecommerce/storefront-qwik-starter/blob/main/src/routes/layout.tsx#L31) where we can get the locale, store it outside of the Qwik context and use it globally within the app. However, I think for better readability of the codebase your approach of passing the locale parameter to every sdk looks better.

@gioboa any thoughts on this matter?

**Comment by gioboa** [comment]
Passing the parameter seems the cleanest solution.
Let's try with this approach

**Comment by gioboa** [comment]
I need to investigate into the problem, I will look at it in few days. You are more then welcome to suggest a solution with a PR

**Comment by prasmalla** [comment]
my approach was a global config variable for currencyCode and use it here
```
const executeRequest = async (options: Options) => {
	const httpResponse = await fetch(
		`${ENV_VARIABLES.VITE_VENDURE_PUBLIC_URL}?currencyCode=${getCurrencyCode()}`,
		options
	);
	return await extractTokenAndData(httpResponse);
};
```

**Comment by gioboa** [comment]
This sounds reasonable to me 👍
Thanks for sharing

**Comment by redaready** [comment]
can we change locale per request in this way? sometimes storefront has a Language Selector component, we need to get locale value from request context(cookie, path, or header) to call api. i think the locale value proprity chain could be 

1. user selection on web interface Language Selector (locale value persisted in cookie ou path, cookie  prefered)
2. user preference setting (stocked on server side,  not in vendure now)
3.  server side app globle deflaut config (as @prasmalla said)
4.  browser language header

**Comment by gioboa** [comment]
Yep, if we change the path with a variable we can do whatever we want

**Comment by prasmalla** [comment]
@redaready we did as you mentioned storing the user selection in a cookie and updating the client-side global config on changes. vendure backend already supports it and there was an [issue here](https://github.com/vendure-ecommerce/vendure/issues/2469) that came up with this approach which has been fixed

**Comment by shubhamchoudhary-2003** [comment]
Hi @gioboa  sir , search in default vendure do not work with the search can we chnage startegy or query or make some multilingual search pugin same error in showing plugin in collection as it use search producxt gql  to display product. , if we can work on it any suggestion or anything from your side

**Comment by gioboa** [comment]
@redaready did you solve this issue? If yes, how did you pass the locale to the vendure calls?

**Comment by shubhamchoudhary-2003** [comment]
Hi sir , the only solution in my opinion is to make a new  search plug-in that support language , I will make plug-in for type sense and try to use to get language in search

**Comment by gioboa** [comment]
I see, you can read few other alternatives [here](https://github.com/wmertens/compiled-i18n/blob/main/docs%2Fqwik.md#client-ui) 👀
[getLocale](https://github.com/QwikDev/qwik/blob/main/packages%2Fqwik%2Fsrc%2Fcore%2Fuse%2Fuse-locale.ts)
